server/service/software_title_icons.go

```diff
@@ -0,0 +1,276 @@
+	}
+	teamIDUint := uint(teamIDUint64)
+
+	decoded := putSoftwareTitleIconRequest{
+		TitleID: uint(titleIDUint64),
+		TeamID:  &teamIDUint,
+	}
+
+	err = r.ParseMultipartForm(512 * units.MiB)
+	if err != nil {
+		return nil, &fleet.BadRequestError{
+			Message:     "failed to parse multipart form",
+			InternalErr: err,
+		}
+	}
+
+	if r.MultipartForm.File["icon"] == nil || len(r.MultipartForm.File["icon"]) == 0 {
+		return nil, &fleet.BadRequestError{
+			Message:     "icon multipart field is required",
+			InternalErr: err,
+		}
+	}
+
+	decoded.File = r.MultipartForm.File["icon"][0]
+
+	return &decoded, nil
+}
+
+func putSoftwareTitleIconEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (fleet.Errorer, error) {
+	req := request.(*putSoftwareTitleIconRequest)
+
+	payload := &fleet.UploadSoftwareTitleIconPayload{
+		TitleID: req.TitleID,
+		TeamID:  *req.TeamID,
+	}
+
+	file, err := req.File.Open()
+	if err != nil {
+		return putSoftwareTitleIconResponse{Err: err}, nil
+	}
+	defer file.Close()
+
+	// ensure icon is png and fits sizing restrictions
+	err = iconValidator(file)
+	if err != nil {
+		return putSoftwareTitleIconResponse{Err: err}, nil
+	}
+
+	tfr, err := fleet.NewTempFileReader(file, nil)
+	if err != nil {
+		return putSoftwareTitleIconResponse{Err: err}, nil
+	}
+	defer tfr.Close()
+	payload.IconFile = tfr
+	payload.Filename = req.File.Filename
+
+	softwareTitleIcon, err := svc.UploadSoftwareTitleIcon(ctx, payload)
+	if err != nil {
+		return putSoftwareTitleIconResponse{Err: err}, nil
+	}
+	iconURL := fmt.Sprintf("/api/latest/fleet/software/titles/%d/icon?team_id=%d", softwareTitleIcon.SoftwareTitleID, softwareTitleIcon.TeamID)
+
+	return putSoftwareTitleIconResponse{
+		IconUrl: iconURL,
+	}, nil
+}
+
+func (svc *Service) UploadSoftwareTitleIcon(ctx context.Context, payload *fleet.UploadSoftwareTitleIconPayload) (fleet.SoftwareTitleIcon, error) {
+	// skipauth: No authorization check needed due to implementation returning
+	// only license error.
+	svc.authz.SkipAuthorization(ctx)
+
+	return fleet.SoftwareTitleIcon{}, fleet.ErrMissingLicense
+}
+
+func iconValidator(file multipart.File) error {
+	config, format, err := image.DecodeConfig(file)
+	if err != nil || format != "png" {
+		return &fleet.BadRequestError{Message: "icon must be a PNG image"}
+	}
+
+	maxWidth, maxHeight := 1024, 1024
+	minWidth, minHeight := 120, 120
+
+	if config.Width > maxWidth || config.Height > maxHeight {
+		return &fleet.BadRequestError{Message: fmt.Sprintf("icon must be no larger than %dx%d pixels", maxWidth, maxHeight)}
+	}
+	if config.Width < minWidth || config.Height < minHeight {
+		return &fleet.BadRequestError{Message: fmt.Sprintf("icon must be at least %dx%d pixels", minWidth, minHeight)}
+	}
+	if config.Width != config.Height {
+		return &fleet.BadRequestError{Message: fmt.Sprintf("icon must be a square image (%dx%d pixels)", config.Width, config.Height)}
+	}
+
+	if _, err := file.Seek(0, 0); err != nil {
```

Comment:
Any reason not to use `.Rewind()` here?

Reply:
I don't think `multipart.File` implements `.Rewind()`?
<img width="668" height="129" alt="Screenshot 2025-08-31 at 7 16 55 PM" src="https://github.com/user-attachments/assets/c84afb0c-bf8e-4b60-b6f9-abd559522536" />

Reply:
Huh, fair enough.